### ISSUE TITLE
🐛 [Bug]:  version v2.52.2 up to v2.52.3 , cors error

### ISSUE THREAD

**Issue by febelery** [gofiber/fiber#2936]
### Bug Description

version v2.52.2 up to v2.52.3 , cors error

version v2.52.2 is correct

### How to Reproduce

1. Go to '....'
3. Do '....'
4. See '....'

### Expected Behavior

cors correct

### Fiber Version

v2.52.3

### Code Snippet (optional)

```go
package main

import "github.com/gofiber/fiber/v2"
import "log"

func main() {
  app := fiber.New()

  	app.Use(cors.New(cors.Config{
		AllowOriginsFunc: func(origin string) bool {
			for _, c := range config.Get[[]string]("cors") {
				if c == origin {
					return true
				}
			}
			return false
		},
	}))

  log.Fatal(app.Listen(":3000"))
}
```

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have checked for existing issues that describe my problem prior to opening this one.
- [X] I understand that improperly formatted bug reports may be closed without explanation.

**Comment by bytesizedwizard** [comment]
I am getting the same issue after upgrading to `2.5.3`. The cors settings seem to be ignored and all requests are blocked. Downgrading to `2.5.2` fixes it.

Just spent the last 3 hours scratching my head an re-evaluating my life choices only to realise it was an error in the package itself 😮‍💨

**Comment by LKaemmerling** [comment]
We are affected with the same issue, after a rollback to 2.5.2 it works again.

**Comment by ReneWerner87** [comment]
@LKaemmerling @bytesizedwizard just to make sure you mean the last release `2.52.3` and not `2.5.x`

can you give us the cors error so that we can fix it as soon as possible 
@febelery 
and also the output of 
```go
app.Use(cors.New(cors.Config{
    AllowOriginsFunc: func(origin string) bool {
        // request origin is important for us
        log.Println("request origin", origin)
        for _, c := range config.Get[[]string]("cors") {
            // your origin is important for us
            log.Println("match origin", c)
            if c == origin {
                return true
            }
        }
        return false
    },
}))
 ```

**Comment by LKaemmerling** [comment]
@ReneWerner87 yes, sorry of course i was talking about 2.52.3. What i can see is that not a single CORS header is added to the response, so regardless of what we configure.

**Comment by ReneWerner87** [comment]
@sixcolors  we should also include the case of the consumers in our unittests so that this no longer happens
@LKaemmerling can you please log the input parameter of the function in the old and new version and give it to us

and also the response headers of the old and new version
possibly also the request origin headers

**Comment by ReneWerner87** [comment]
@LKaemmerling can you give the request origin and also the response headers in the old and new versions for you

we need the possibility to recreate the problem on our site in order to correct it

it is also good to know that you are not using the function, then it probably has something to do with the following process
at first i thought it was something specific to the function

**Comment by LKaemmerling** [comment]
@ReneWerner87 working (with 2.52.2):
```
(copied from browser)
Req:
curl 'https://api.our-staging-environment.works/v1/datacenters?page=1&per_page=50' \
  -H 'accept: application/json, text/plain, */*' \
  -H 'accept-language: de-DE,de;q=0.9,en-DE;q=0.8,en;q=0.7,en-US;q=0.6,fr;q=0.5' \
  -H 'authorization: Bearer yoOneM2yk4CKfqrwyuYz8erxE8ZlUQp0HcuU3ku2s3LHK8sS672pm800fXJ69DI4' \
  -H 'content-type: application/json' \
  -H 'origin: https://our-staging-environment.works' \
  -H 'referer: https://our-staging-environment.works/' \
  -H 'sec-ch-ua: "Google Chrome";v="123", "Not:A-Brand";v="8", "Chromium";v="123"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-site' \
  -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36'
  
  
  Resp:
  Access-Control-Allow-Origin:
*
Access-Control-Expose-Headers:
Link,X-Correlation-ID
Content-Length:
6538
Content-Type:
application/json
Date:
Tue, 26 Mar 2024 10:15:28 GMT
Link:
<https://staging-environment.work/v1/datacenters?page=1&per_page=50>; rel="last"
Ratelimit-Limit:
3600
Ratelimit-Remaining:
3596
Ratelimit-Reset:
1711448132
Strict-Transport-Security:
max-age=31536000; includeSubDomains
Traceparent:
00-fadc03b7196b8c11be29b53baeabd8ad-c37a7132aefb1a01-01
Uber-Trace-Id:
fadc03b7196b8c11be29b53baeabd8ad:c37a7132aefb1a01:0:1
Vary:
Origin
X-Correlation-Id:
fadc03b7196b8c11be29b53baeabd8ad

  ```
  
  Broken: (v2.52.3)
  ```
  Req is the same

Response:
Content-Length:
6538
Content-Type:
application/json
Date:
Tue, 26 Mar 2024 10:22:11 GMT
Link:
<https://staging-environment.work/v1/datacenters?page=1&per_page=50>; rel="last"
Ratelimit-Limit:
3600
Ratelimit-Remaining:
3593
Ratelimit-Reset:
1711448538
Strict-Transport-Security:
max-age=31536000; includeSubDomains
Traceparent:
00-781e9af000c65354efd7695d6dc5595b-9bc3d5996294ead6-01
Uber-Trace-Id:
781e9af000c65354efd7695d6dc5595b:9bc3d5996294ead6:0:1
X-Correlation-Id:
781e9af000c65354efd7695d6dc5595b
```

So you can see that the CORS headers are missing.

**Comment by ReneWerner87** [comment]
related issue https://github.com/gofiber/fiber/issues/2920
![image](https://github.com/gofiber/fiber/assets/7063188/ba46eb5f-54ed-485e-92c9-f0f5f6393186)

releated fix https://github.com/gofiber/fiber/pull/2921
![image](https://github.com/gofiber/fiber/assets/7063188/ae278a28-7180-466d-868a-b41a07745683)

**Comment by ReneWerner87** [comment]
believe this check should only be added to the options request and not to the actual request of the action

**Comment by sixcolors** [comment]
> @sixcolors this is the problem
> 
> the second condition
> 
> https://github.com/gofiber/fiber/blob/7ba02c14cf53c5a5f8171a0e4eb125ef4615f7c0/middleware/cors/cors.go#L174

That’s not a CORS request. The middleware was swinging headers when it should not have before.

**Comment by ReneWerner87** [comment]
we could confirm that our new logic, which is closer to the standard, is unfortunately not decapsulated and also affects the non-preflight request

we will adjust this by the end of the day or tomorrow morning and then release in a timely manner

until then I would suggest skipping the last release

**Comment by sixcolors** [comment]
@febelery @LKaemmerling and @bytesizedwizard are you able to verify if #2937 fixes CORS in your dev envs?

**Comment by ReneWerner87** [comment]
@LKaemmerling @febelery @bytesizedwizard 

can you pls briefly check if everything fits with the following commit before I release and tag it

go.mod
```gomod
require github.com/gofiber/fiber/v2 v2.52.4-0.20240326212242-a6f4c133bc74
```

**Comment by LKaemmerling** [comment]
@ReneWerner87 @sixcolors thank you a lot! With your recent changes i can confirm it now works again as we would expect it.

**Comment by ReneWerner87** [comment]
thanks, but wait until the other two confirm, then I will release

**Comment by luisgarciaalanis** [comment]
Trying to repro this issue, but the CORS AllowOriginsFunc function is never called, any ideas on how to trigger it?
```
func main() {
	engine := html.New("./views", ".html")
	allowedOrigins := []string{"http://garcia.tv"}
	app := fiber.New(fiber.Config{
		Views: engine,
	})

	fmt.Println("Setting up CORS")
	app.Use(cors.New(cors.Config{
		AllowMethods: "GET,POST,PUT,OPTIONS",
		AllowHeaders: "Origin,Content-Type,Accept,Content-Length,Accept-Language,Accept-Encoding,Connection,Access-Control-Allow-Origin",
		AllowOriginsFunc: func(origin string) bool {
			// request origin is important for us
			log.Println("request origin", origin)
			for _, c := range allowedOrigins {
				// your origin is important for us
				log.Println("match origin", c)
				if c == origin {
					return true
				}
			}
			return false
		},
	}))

	app.Get("/", func(c *fiber.Ctx) error {
		// Render index
		return c.Render("index", fiber.Map{
			"Title": "Hello, World!",
		})
	})

	log.Fatal(app.Listen(":3000"))
}
```

**Comment by luisgarciaalanis** [comment]
Can this be forward ported to v3?

**Comment by febelery** [comment]
> @LKaemmerling @febelery @bytesizedwizard
> 
> can you pls briefly check if everything fits with the following commit before I release and tag it
> 
> go.mod
> 
> ```
> require github.com/gofiber/fiber/v2 v2.52.4-0.20240326212242-a6f4c133bc74
> ```

work for me, thanks

**Comment by ReneWerner87** [comment]
solved with [v2.52.4](https://github.com/gofiber/fiber/releases/tag/v2.52.4)

> Can this be forward ported to v3?

we will do this

**Comment by sixcolors** [comment]
> solved with [v2.52.4](https://github.com/gofiber/fiber/releases/tag/v2.52.4)
> 
> > Can this be forward ported to v3?
> 
> we will do this

@ReneWerner87 #2937 is the v3 fix.
